mlflow/utils/_capture_modules.py

```diff
@@ -95,6 +95,26 @@ def __exit__(self, *_, **__):
         importlib.import_module = self.original_import_module
 
 
+class _CaptureImportedModulesForHF(_CaptureImportedModules):
+    """
+    A context manager to capture imported modules by temporarily applying a patch to
+    `builtins.__import__` and `importlib.import_module`.
+    Used for 'transformers' flavor only.
+    """
+
+    def __init__(self, module_to_throw):
+        super().__init__()
+        self.module_to_throw = module_to_throw
+
+    def _wrap_package(self, name):
+        if name == self.module_to_throw or name.startswith(f"{self.module_to_throw}."):
+            raise ImportError(f"Disabled package {name}")
```

Comment:
Instead of `ImportError`, can we create a custom error?

```python
class NotAllowed(Exception):
    def __init__(self, module):
        super().__init__(f"{module} is not allowed to import")
```

This helps simplify try-catch:

```python
try:
     ...
except NotAllowed:
     pass
```

Reply:
When we load the model back with transformers, it internally raise a RuntimeError here:
<img width="1127" alt="image" src="https://github.com/mlflow/mlflow/assets/82044803/d97cc391-00a8-4660-925b-952780cbffa6">
So even if the original error is ImportError or this new exception, we can't catch it :(

Reply:
I see, can we catch `RuntimeError` and check the `__cause__` attribute?

Reply:
For example,

```python
class MyExcpetion(Exception):
    def __init__(self):
        super().__init__("MyExcpetion")


def throw():
    try:
        raise MyExcpetion
    except MyExcpetion as e:
        raise RuntimeError("runtime error") from e


try:
    throw()
except RuntimeError as e:
    print(e, e.__cause__, type(e.__cause__))

```